devtools/src/arcs-hcr-list.js

```diff
@@ -0,0 +1,297 @@
+        <div divider></div>
+        <div class="toggleicon">
+          <label class="switch">
+            <input type="checkbox" on-click="_loadServer" id="toggle">
+            <span class="slider round"></span>
+          </label>
+          <span class="toggletext">
+            Hot Code Reload
+          </span>
+        </div>
+      </div>
+    </header>
+
+    <div main>
+      <div class="title">Reloadable Particles</div>
+      <div class="content">
+        <template is="dom-repeat" items="{{reloadableParticles}}">
+          <object-explorer log-id$="[[item.implFile]]" object="{{item}}">
+            <span name>[[item.name]]</span>
+            <span id>[[item.id]]</span>
+          </object-explorer>
+        </template>
+        <template is="dom-if" if="{{!item.items.length}}">
+          <div class="empty-label">No Particles</div>
+        </template>
+      </div>
+    </div>`;
+  }
+
+  static get is() { return 'arcs-hcr-list'; }
+
+  constructor() {
+    super();
+    this.websocket = null;
+    this.particleSources = new Set();
+    this.arcIdsToKeepFresh = [];
+
+    this.pathToTimeLog = new Map();
+    this.reloadableParticles = [];
+    this.allParticlesInArc = [];
+  }
+
+  onRawMessageBundle(messages) {
+    for (const msg of messages) {
+      switch (msg.messageType) {
+        case 'arc-available': {
+          if (!msg.messageBody.speculative) {
+            const id = msg.meta.arcId.toString();
+            this.arcIdsToKeepFresh.push(id);
+          }
+          break;
+        }
+        case 'watch-particle-sources': {
+          const before = this.particleSources.size;
+          for (const particleSource of msg.messageBody) {
+            this.particleSources.add(particleSource);
+          }
+          if (this.particleSources.size > before && this.websocket != null) {
+            const files = JSON.stringify([...this.particleSources]);
+            this.websocket.send(files);
+          } 
+          break;
+        }
+      }
+    }
+  }
+
+  onMessageBundle(messages) {
+    for (const msg of messages) {
+      switch (msg.messageType) {
+        case 'arc-selected':
+          this.reloadableParticles = [];
+          this.allParticlesInArc = [];
+          break;
+
+        case 'PecLog': {
+          const m = msg.messageBody;
+          switch (m.name) {
+            case 'InstantiateParticle': {
+              const particleId = m.pecMsgBody.id.toString();
+              const spec = m.pecMsgBody.spec;
+
+              this.allParticlesInArc.push({id: particleId, spec: spec});
+              this._displayLog(particleId, spec);
+            }
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  _forceReload() {
+    this.reloadableParticles.forEach(particle => {
+      this._updateLog(particle.implFile);
+      this.arcIdsToKeepFresh.forEach(id => {
+        this.send({
+            messageType: 'particle-reload',
```

Comment:
Interesting - so the "reload" button in the UI sends a product of [list of particle] x [arc ids], requesting to reload each combination? I'm not convinced we need this, would you be fine with removing it for now and if we learn we do want a global "reload everything" button, then we add something like this?

Reply:
Yes I don't really mind removing it for now (currently we probably also have no reason to reload a particle when nothing is changed). Removed! :)